### PR TITLE
Add `quantum_fisher_information_matrix` to `quantum_info.metrics`

### PR DESCRIPTION
**Pull request by renatomello**
Checklist:
- [x] Reviewers confirm new code works as expected.
- [x] Tests are passing.
- [x] Coverage does not decrease.
- [x] Documentation is updated.

### CODE REVIEW

File: src/qibo/quantum_info/metrics.py [hunk 1]
```diff
@@ -845,6 +845,27 @@ def frame_potential(
...
+    if parameters is None:
+        parameters = circuit.get_parameters()
+        parameters = backend.cast(parameters, dtype=float)
+
+    jacobian = backend.calculate_jacobian_matrix(circuit, parameters, return_complex)
+
+    original_parameters = circuit.get_parameters()
+    circuit.set_parameters(parameters)
+    state = backend.execute_circuit(circuit).state()
+    circuit.set_parameters(original_parameters)
```
Comment (alecandido) [inline src/qibo/quantum_info/metrics.py:862]: Instead of doing this, you may just decide to copy the circuit in case the `parameters` argument is specified
```suggestion
    if parameters is not None:
        circuit = copy.deepcopy(circuit)
        circuit.set_parameters(parameters)
    else:
        parameters = backend.cast(circuit.get_parameters(), dtype=float)

    jacobian = backend.calculate_jacobian_matrix(circuit, parameters, return_complex)
    state = backend.execute_circuit(circuit).state()
```

To me, it seems conceptually simpler.

**Review by alecandido — COMMENTED** [review]
I'm essentially fine with it, just a few minor comments.

**Comment by alecandido** [comment]
Ah, the tests are not passing (because of the linter), that should be fixed (of course - but the CI is already there to remind you...).

**Comment by alecandido** [comment]
Btw, @MatteoRobbiati, pay attention to this PR, since it will affect the TensorFlow/Pytorch backend migration to Qiboml.

File: src/qibo/backends/pytorch.py [hunk 2]
```diff
@@ -239,6 +239,17 @@ def calculate_matrix_power(
         copied = super().calculate_matrix_power(copied, power, precision_singularity)
         return self.cast(copied, dtype=copied.dtype)
 
+    def calculate_jacobian_matrix(self, circuit, params, return_complex: bool = True):
+        def func(params):
+            """torch requires object(s) to be wrapped in a function."""
+            circuit.set_parameters(params)
+            state = self.execute_circuit(circuit).state()
+            if return_complex:
```
[the remark this answers is not in the document]
Comment (renatomello) [reply src/qibo/backends/pytorch.py:247]: Locally it is, at least for the examples I tested.

File: src/qibo/backends/abstract.py [hunk 3]
```diff
@@ -376,6 +376,13 @@ def calculate_singular_value_decomposition(self, matrix):  # pragma: no cover
         """Calculate the Singular Value Decomposition of ``matrix``."""
         raise_error(NotImplementedError)
 
+    @abc.abstractmethod
+    def calculate_jacobian_matrix(
```
[the remark this answers is not in the document]
Comment (renatomello) [reply src/qibo/backends/abstract.py:380]: I was following the same name pattern of other backend methods like `calculate_eigenvalues`, `calculate_matrix_power`, etc.

Maybe it's worth to open another PR to rename them all.

[the remark this answers is not in the document]
Comment (alecandido) [reply src/qibo/backends/abstract.py:380]: > I was following the same name pattern of other backend methods like `calculate_eigenvalues`, `calculate_matrix_power`, etc.

Ok, fine, keep going.

> Maybe it's worth opening another PR to rename them all.

In case you consider it meaningful, let's open an issue for that, and accurately schedule for the future.
Bulk changing backend functions (even if not the most frequently used, like `execute_circuit`) is definitely a breaking change. Better to plan it carefully.

[the remark this answers is not in the document]
Comment (renatomello) [reply src/qibo/backends/abstract.py:380]: I don't consider it meaningful TBH

[the remark this answers is not in the document]
Comment (alecandido) [reply src/qibo/backends/abstract.py:380]: > I don't consider it meaningful TBH

I would have chosen different names. In general, the simplest non-ambiguous ones. But I prefer avoiding breaking.

To be fair, I personally consider `jacobian` different from `calculate_eigenvalues`, since the Jacobian requires a dedicated execution (while the other one is about results manipulation).
But this means to me that the Jacobian will be propagated to `qibo-core`, while `calculate_eigenvalues` will be split out the `qibo-core` backends.

Until that time comes, do whatever you wish.

File: doc/source/api-reference/qibo.rst [hunk 4]
```diff
@@ -1951,6 +1951,12 @@ Frame Potential
 .. autofunction:: qibo.quantum_info.frame_potential
 
 
+Quantum Fisher information matrix
```
Comment (renatomello) [inline doc/source/api-reference/qibo.rst:1954]:
```suggestion
Quantum Fisher Information Matrix
```

**Review by alecandido — APPROVED** [review]

File: src/qibo/backends/pytorch.py [hunk 5]
```diff
@@ -239,6 +239,21 @@ def calculate_matrix_power(
         copied = super().calculate_matrix_power(copied, power, precision_singularity)
         return self.cast(copied, dtype=copied.dtype)
 
+    def calculate_jacobian_matrix(
+        self, circuit, parameters=None, initial_state=None, return_complex: bool = True
+    ):
```
Comment (BrunoLiegiBastonLiegi) [inline src/qibo/backends/pytorch.py:244]: Just to make sure, this is the jacobian of the state after execution with respect to the circuit parameters, right?

Comment (renatomello) [reply src/qibo/backends/pytorch.py:244]: Yes. The execution per se is not strictly necessary, we do it because that's how we get the gradients. But it's jacobian of a state w.r.t. the phases of the gates that creates said state.
